File: guests/swift/entry.swift
// The entry scene, Swift port — guests/rust/entry.rs,
// tools/scenes/entry.steps.

import Foundation

let app = KayaApp()

let (status, field, add, todos) = app.build {
    tx -> (KayaSignal, KayaWidget, KayaWidget, KayaCollection) in
    let status = tx.signal(.str("no todos"))
    let todos = tx.collection()

    let (root, field, add) = tx.column { root -> (KayaWidget, KayaWidget, KayaWidget) in
        let field = tx.entry()  // entry#0
        let add = tx.button("add")  // button#0
        tx.label(bind: status)  // label#0
        tx.each(todos) { t in
            t.label(KayaField<String>(index: 0))
        }
        return (root, field, add)
    }
    tx.mount(root)
    return (status, field, add, todos)
}

var draft = ""
app.onChange(field) { _, text in
    draft = text
}
app.onClick(add) { tx in
    if draft.isEmpty {
        tx.write(status, .str("nothing to add, \(tx.count(todos)) total"))
        return
    }
    tx.insertFresh(todos, .str(draft))
    let total = tx.count(todos)
    tx.write(status, .str("added \(draft), \(total) total"))
    tx.clear(field)
    tx.focus(field)
}

app.run()

File: guests/ocaml/entry.ml
(* The entry scene, OCaml port — guests/rust/entry.rs,
   tools/scenes/entry.steps. *)

open Kaya_wire
open Kaya_app

let () =
  let app = Kaya_app.create () in

  let status, field, add, todos =
    build app (fun () ->
       let status = signal (Str "no todos") in
       let todos = collection () in

       let field = entry () in
       let add = button ~text:"add" () in

       let root =
         column
           [
             w field (* entry#0 *);
             w add (* button#0 *);
             label ~bind:status (* label#0 *);
             each todos (fun () -> Tpl.(label ~bind_field:element ()));
           ]
           ()
       in
       mount root;
       (status, field, add, todos))
  in

  let draft = ref "" in
  on_change app field (fun text -> draft := text);
  on_click app add (fun () ->
     let d = !draft in
     if d = "" then
       let total = count todos in
       write status (Str (Printf.sprintf "nothing to add, %d total" total))
     else begin
       ignore (insert_fresh todos (Str d));
       let total = count todos in
       write status (Str (Printf.sprintf "added %s, %d total" d total));
       (* The clear comes back as text_changed "", so the fold empties
          the draft. *)
       clear field;
       focus field
     end);

  exit (run app)
